# Using ROBOKOP's Quick Question Answering Service

The expand, enrich, and similarity services each offer bite-sized approaches to interacting with ROBOKOP.  The ROBOKOP "quick" service offers a slightly more complex approach to answering questions.   We'll be using the following function to call the quick service.  Note that calling the service requires posting a question.  We'll discuss the format of the question below.

In [1]:
robokop_server='robokop.renci.org'

In [2]:
import requests

def quick(question,max_results=None,output_format=None):
    url=f'http://{robokop_server}:80/api/simple/quick/'
    if max_results is not None:
        url += f'?max_results={max_results}'
    if output_format is not None:
        j = '&' if '?' in url else '?'
        url += f'{j}output_format={output_format}'
    response = requests.post(url,json=question)
    print( f"Return Status: {response.status_code}" )
    if response.status_code == 200:
        return response.json()
    return response

## Question Format and Basic Usage

The question is a python dictionary.  It takes a key `machine_question`, which is a dictionary containing a list of `nodes` and a list of `edges`.   Each node object needs an integer `id`.   Any node can also have a `curie` specifying that node.  The edges define connections between the nodes using the identifiers of the nodes as source and targets. The following function shows how to construct a single-hop question.  It takes a specified node of `type1` and looks for any node of `type2`.

In [3]:
def make_one_step_question(type1, id1, type2,rebuild = None):
    question = {
                'machine_question': {
                    'nodes': [
                        {
                            'id': 'n0',
                            'curie': id1,
                            'type': type1
                        },
                        {
                            'id': 'n1',
                            'type': type2
                        }
                    ],
                    'edges': [
                        {
                            'id': 'e0',
                            'source_id': 'n0',
                            'target_id': 'n1'
                        }
                    ]
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild'] = 'True'
    return question

Here, we will specify a one-hop question asking for the phenotypes associated with Fanconi Anemia (MONDO:0019391).  We first construct the question then use it to call the quick service.  In fact, this is how the expand function is implemented internally.  The output is returned in the KG standard answer format.

In [4]:
q = make_one_step_question('disease','MONDO:0019391','phenotypic_feature')
r = quick(q)

Return Status: 200


In [5]:
# Uncomment to see a big long answer
import json
#print( json.dumps(r, indent=4))

The output can be simplified with a function like this:

In [6]:
import pandas as pd

def parse_answer(returnanswer):
    #First, parse out the parts of the kg that we want, names and types
    kg_node_names = { n['id']: n['name'] if 'name' in n else n['id'] for n in returnanswer['knowledge_graph']['nodes'] }
    kg_edge_types = { e['id']: e['type'] for e in returnanswer['knowledge_graph']['edges']}
    kg_edge_sources = { e['id']: e['edge_source'] for e in returnanswer['knowledge_graph']['edges']}
    #now put the answers into a table, looking up node/edge info from the above dicts
    #because of the question, we know that we're interested in the binding for "n1" and "e0"
    #this is quick-n-dirty, so we're also limiting ourselves to the first edge, even if there is more than one
    answers = [ {"result_id": answer['node_bindings']['n1'], 
                 "result_name": kg_node_names[answer['node_bindings']['n1']], #if 'name' in node else node['id'], 
                 "type":        kg_edge_types[answer['edge_bindings']['e0'][0]],
                 "source":      kg_edge_sources[answer['edge_bindings']['e0'][0]],
                 "score" :      answer['score']}
              for answer in returnanswer['answers']]
    return pd.DataFrame(answers)


In [7]:
parse_answer(r)

result_id                                     result_name     score  \
0    HP:0005505                               Refractory anemia  1.619721   
1    HP:0100615                                Ovarian neoplasm  1.619721   
2    HP:0002488                                  Acute leukemia  1.619721   
3    HP:0002860                         Squamous cell carcinoma  1.619721   
4    HP:0004430                Severe combined immunodeficiency  1.619721   
5    HP:0002721                                Immunodeficiency  1.619721   
6    HP:0001402                        Hepatocellular carcinoma  1.619721   
7    HP:0000286                                      Epicanthus  1.619721   
8    HP:0003002                                Breast carcinoma  1.619721   
9    HP:0004377                          Hematological neoplasm  1.619721   
10   HP:0007354                   Amyotrophic lateral sclerosis  1.619721   
11   HP:0008064                                      Ichthyosis  1.619721   
12   HP:0006721                    Acute lymphoblastic leukemia  1.619721   
13   HP:0003254                       Abnormality of DNA repair  1.619721   
14   HP:0001915                                 Aplastic anemia  1.619721   
15   HP:0000824                       Growth hormone deficiency  1.619721   
16   HP:0004808                          Acute myeloid leukemia  1.619721   
17   HP:0005560                 Imbalanced hemoglobin synthesis  1.619721   
18   HP:0100806                                          Sepsis  1.619721   
19   HP:0005510                    Transient erythroblastopenia  1.619721   
20   HP:0001972                               Macrocytic anemia  1.619721   
21   HP:0001251                                          Ataxia  1.619721   
22   HP:0011018                   Abnormality of the cell cycle  1.619721   
23   HP:0004431                           Complement deficiency  1.619721   
24   HP:0004810                   Congenital hypoplastic anemia  1.619721   
25   HP:0000789                                     Infertility  1.619721   
26   HP:0000953                   Hyperpigmentation of the skin  1.619721   
27   HP:0001900                            Increased hemoglobin  1.619721   
28   HP:0001908                              Hypoplastic anemia  1.619721   
29   HP:0000135                                    Hypogonadism  1.619721   
..          ...                                             ...       ...   
220  HP:0001053                      Hypopigmented skin patches  0.517653   
221  HP:0001631                            Atrial septal defect  0.517653   
222  HP:0002823                 Abnormality of femur morphology  0.517653   
223  HP:0005344              Abnormal carotid artery morphology  0.517653   
224  HP:0004349                    Reduced bone mineral density  0.517653   
225  HP:0000218                                     High palate  0.517653   
226  HP:0000486                                      Strabismus  0.517653   
227  HP:0008053                  Aplasia/Hypoplasia of the iris  0.517653   
228  HP:0000813                               Bicornuate uterus  0.517653   
229  HP:0001639                     Hypertrophic cardiomyopathy  0.517653   
230  HP:0100026                      Arteriovenous malformation  0.517653   
231  HP:0000864  Abnormality of the hypothalamus-pituitary axis  0.517653   
232  HP:0007565                     Multiple cafe-au-lait spots  0.517653   
233  HP:0008572                       External ear malformation  0.517653   
234  HP:0002007                                 Frontal bossing  0.517653   
235  HP:0004209                  Clinodactyly of the 5th finger  0.517653   
236  HP:0000957                               Cafe-au-lait spot  0.517653   
237  HP:0002023                                    Anal atresia  0.517653   
238  HP:0001903                                          Anemia  0.517653   
239  HP:0100760                                Clubbing of toes  0.517653   
24

## Multi-step questions

It's straightforward to generalize this linear query to an N-item path.  This function constructs a question from a list of types and a list of identifiers.  The types are the types of nodes that will be traversed along the path, and the identifiers represent fixed elements in the path.  The length of types and ids should be equal, and free nodes should specify an id of None:

In [8]:
def make_N_step_question(types,curies,rebuild = None):
    question = {
                'machine_question': {
                    'nodes': [],
                    'edges': []
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild'] = 'True'
    ei = 0
    for i,t in enumerate(types):
        newnode = {'id': f'n{i}', 'type': t}
        if curies[i] is not None:
            newnode['curie'] = curies[i]
        question['machine_question']['nodes'].append(newnode)
        if i > 0:
            question['machine_question']['edges'].append( {'id': f'e{ei}','source_id': f'n{i-1}', 'target_id': f'n{i}'})
            ei += 1
    return question

We can recapitulate our previous question with this new function like this:

In [9]:
newq = make_N_step_question(['disease','phenotypic_feature'],['MONDO:0019391',None])
q == newq

True

Now we could expand to a longer query.  The following question will start at the disease MONDO:0019391, go to a gene, and from there to a biological process or activity.

In [10]:
two_step_question = make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,None])

In [11]:
two_step_answer = quick(two_step_question)

Return Status: 200


We can extract the node names along the paths with a simple function:

In [12]:
def extract_node_names(returnanswer):
    #make a dict from identifiers (like MONDO:0019391) to names (like "Fanconi Anemia")
    kg_node_names = { n['id']: n['name'] if 'name' in n else n['id'] for n in returnanswer['knowledge_graph']['nodes'] }
    #For every answer, make a dict from the query node ids (like n0) to the node name (like Fanconia Anemia)
    nodes = [
                {q_node: kg_node_names[ q_id ] for q_node, q_id in answer['node_bindings'].items()}
                for answer in returnanswer['answers']
            ]
    #Turn this list of dicts into a table
    return pd.DataFrame(nodes)

In [13]:
two_step_nodes = extract_node_names(two_step_answer)
two_step_nodes

n0      n1                                                 n2
0    Fanconi anemia   BRIP1                        response to toxic substance
1    Fanconi anemia   FANCG                              response to radiation
2    Fanconi anemia  RAD51C                                  DNA recombination
3    Fanconi anemia   FANCA                   regulation of cell proliferation
4    Fanconi anemia   FANCA                           female gonad development
5    Fanconi anemia   FANCL                   regulation of cell proliferation
6    Fanconi anemia   BRCA2                           female gonad development
7    Fanconi anemia   BRCA2                 inner cell mass cell proliferation
8    Fanconi anemia   BRCA2                                        hemopoiesis
9    Fanconi anemia   PALB2                 inner cell mass cell proliferation
10   Fanconi anemia   PALB2                         animal organ morphogenesis
11   Fanconi anemia   RAD51                                  DNA recombination
12   Fanconi anemia   ERCC4                                     response to UV
13   Fanconi anemia  MAD2L2           positive regulation of isotype switching
14   Fanconi anemia   ERCC4                     endodeoxyribonuclease activity
15   Fanconi anemia  MAD2L2                                      cell division
16   Fanconi anemia   FANCF                             protein ubiquitination
17   Fanconi anemia     MX1                                    membrane fusion
18   Fanconi anemia     TNF                 endothelial cell apoptotic process
19   Fanconi anemia     TNF               positive regulation of cell adhesion
20   Fanconi anemia     TNF                         microglial cell activation
21   Fanconi anemia     MX1                                   defense response
22   Fanconi anemia     TNF                              inflammatory response
23   Fanconi anemia     TNF                               astrocyte activation
24   Fanconi anemia     TNF                         osteoclast differentiation
25   Fanconi anemia     TNF                  extracellular matrix organization
26   Fanconi anemia     TNF                            humoral immune response
27   Fanconi anemia     MX1                             innate immune response
28   Fanconi anemia     TNF  epithelial cell proliferation involved in sali...
29   Fanconi anemia     TNF                     leukocyte tethering or rolling
..              ...     ...                                                ...
220  Fanconi anemia   RAD51                          identical protein binding
221  Fanconi anemia   ERCC4   transcription-coupled nucleotide-excision repair
222  Fanconi anemia   ERCC4  single-stranded DNA endodeoxyribonuclease acti...
223  Fanconi anemia   ERCC4                            cellular response to UV
224  Fanconi anemia  MAD2L2  RNA polymerase II activating transcription fac...
225  Fanconi anemia   RAD51                        double-stranded DNA binding
226  Fanconi anemia   RAD51                                     enzyme binding
227  Fanconi anemia  MAD2L2   negative regulation of protein catabolic process
228  Fanconi anemia   RAD51  mitotic recombination-dependent replication fo...
229  Fanconi anemia   RAD51      single-stranded DNA-dependent ATPase activity
230  Fanconi anemia   ERCC4                                         DNA repair
231  Fanconi anemia   ERCC4  nucleotide-excision repair involved in interst...
232  Fanconi anemia   ERCC4                                        DNA binding
233  Fanconi anemia   RAD51          DNA unwinding involved in DNA replication
234  Fanconi anemia  MAD2L2  negative regulation of canonical Wnt signaling...
235  Fanconi anemia  MAD2L2  positive regulation of double-strand break rep...
236  Fanconi anemia   ERCC4                         protein N-terminus binding
237  Fanconi anemia   RAD51                                  chromatin binding
238  Fanconi anemia   ERCC4                      telomerase inhibit

## Output Format

### Version 0.9, output_format=message

The results above are returned in v0.9 of the Translator standard for graph answers.  In this format, one graph is returned, containing all of the nodes and edges required for any of the answers.  Then the list of answers is given, binding nodes and edges in the question definition to nodes and edges in the shared knowledge graph.  This allows greater efficiency and specificity in relating outputs to question parameters.

The response has 3 keys returning
1. The json formatted question that was asked 
2. The shared knowledge graph 
3. The answers as defined on that graph

In [29]:
print(two_step_answer.keys())

dict_keys(['question_graph', 'knowledge_graph', 'answers'])


The knowledge graph consists of nodes (containing all the information about the entity) and edges (containing information about the relationship between entities).

In [30]:
print(two_step_answer['knowledge_graph'].keys())

dict_keys(['nodes', 'edges'])


Here's what the first 2 nodes and the first edge look like:

In [31]:
print(json.dumps(two_step_answer['knowledge_graph']['nodes'][:2],indent=4))

[
    {
        "congenital abnormality": true,
        "name": "Fanconi anemia",
        "nutritional or metabolic disease": true,
        "id": "MONDO:0019391",
        "equivalent_identifiers": [
            "MESH:D005199",
            "MEDDRA:10016218",
            "MONDO:0019391",
            "MEDDRA:10055206",
            "UMLS:C0015625",
            "ORPHANET:84",
            "DOID:13636"
        ],
        "type": "disease",
        "rare disease": true,
        "syndromic disease": true
    },
    {
        "name": "UBE2T",
        "location": "1q32.1",
        "locus_group": "protein-coding gene",
        "gene_family": [
            "Ubiquitin conjugating enzymes E2",
            "FA complementation groups"
        ],
        "chromosome": "1",
        "id": "HGNC:25009",
        "equivalent_identifiers": [
            "UniProtKB:Q9NPD8",
            "NCBIGene:29089",
            "HGNC:25009",
            "ENSEMBL:ENSG00000077152",
            "HGNC.SYMBOL:UBE2T",
          

In [33]:
print(json.dumps(two_step_answer['knowledge_graph']['edges'][1],indent=4))

{
    "relation": "PHAROS:gene_involved",
    "target_id": "HGNC:25009",
    "edge_source": "pharos.disease_get_gene",
    "publications": [],
    "id": "4257632",
    "predicate_id": "NCIT:R176",
    "source_database": "pharos",
    "source_id": "MONDO:0019391",
    "type": "disease_to_gene_association",
    "ctime": 1543738834.740142,
    "relation_label": "gene_involved",
    "weight": 0.25832639515782074
}


Here's what the first answer looks like:

In [34]:
print(json.dumps(two_step_answer['answers'][0],indent=4))

{
    "node_bindings": {
        "n0": "MONDO:0019391",
        "n1": "HGNC:20473",
        "n2": "GO:0009636"
    },
    "edge_bindings": {
        "e0": [
            "4257531",
            "6938566",
            "6938600"
        ],
        "e1": [
            "4294578",
            "1094816"
        ]
    },
    "score": 0.3959587237470695
}


The answers are ranked by score, so this is the highest ranking answer, and the score is given.  The `node_bindings` relate the identifiers of the questions (`n0`, `n1`, `n2`) to the identifiers in the knowledge graph (`MONDO:0019391`, `HGNC:20473`, `GO:0009636`) and we could use these identifiers to look up further information about those nodes in the graph.

Similarly, the `e0` edge (connecting `n0` and `n1` in the question), is mapped to several possible edges in the graph, and we could look up the edge information using the identifier.  For instance, we can find the first edge between for `e0`:

In [41]:
scores = [a['score'] for a in two_step_answer['answers']]
scores

[0.3959587237470695,
 0.39562714985581743,
 0.3930911628130927,
 0.38680121546632074,
 0.38680121546632074,
 0.38237448152146924,
 0.37742069307012543,
 0.37742069307012543,
 0.37742069307012543,
 0.35967233858255104,
 0.35967233858255104,
 0.3328828490620355,
 0.3328828490620355,
 0.3328828490620355,
 0.3328828490620355,
 0.3328828490620355,
 0.330896351413593,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.22921193146495913,
 0.22921193146495913,
 0.22921193146495913,
 0.22921193146495913,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 

In [35]:
my_edge = list(filter(lambda x: x['id'] == "4257531", two_step_answer['knowledge_graph']['edges']))[0]
print(json.dumps(my_edge,indent=4))

{
    "relation": "PHAROS:gene_involved",
    "target_id": "HGNC:20473",
    "edge_source": "pharos.disease_get_gene",
    "publications": [],
    "id": "4257531",
    "predicate_id": "NCIT:R176",
    "source_database": "pharos",
    "source_id": "MONDO:0019391",
    "type": "disease_to_gene_association",
    "ctime": 1543738834.73799,
    "relation_label": "gene_involved",
    "weight": 0.25832639515782074
}


### v0.8, output_format=dense

For backwards compatibility, output can also be returned in v0.8 of the KG format.  To return output in this format, pass the argument `output_format=dense` to the url.   The `output_format` defaults to `message` which specifies v0.9.

In [36]:
two_step_answer_dense = quick(two_step_question,output_format='dense')

Return Status: 200


In [38]:
two_step_answer_dense.keys()

dict_keys(['datetime', 'id', 'message', 'response_code', 'result_list'])

In [42]:
two_step_answer_dense['result_list'][0]

{'confidence': 0.3959587237470695,
 'id': '5939821d-f96e-442c-ab37-059ffe1087e4',
 'result_graph': {'edge_list': [{'confidence': 0.25832639515782074,
    'num_publications': None,
    'provided_by': 'pharos.disease_get_gene',
    'publications': '',
    'source_id': 'MONDO:0019391',
    'target_id': 'HGNC:20473',
    'type': 'disease_to_gene_association'},
   {'confidence': 0.4071474314830641,
    'num_publications': None,
    'provided_by': 'biolink.disease_get_gene',
    'publications': 'PMID:20639400',
    'source_id': 'MONDO:0019391',
    'target_id': 'HGNC:20473',
    'type': 'biomarker_for'},
   {'confidence': 0.9998780282217139,
    'num_publications': None,
    'provided_by': 'biolink.disease_get_gene',
    'publications': 'PMID:25980754,PMID:19127258,PMID:16116423,PMID:16153896,PMID:16116424,PMID:27498913,PMID:22006311,PMID:24556621,PMID:26689913,PMID:26681312,PMID:17033622,PMID:26720728,PMID:25186627,PMID:25452441,PMID:26921362,PMID:21964575,PMID:26786923,PMID:21127055,PMID:2

## Specifying multiple fixed nodes

The examples above have all started from one known node and expanded from it, sometimes in multiple steps.  It's also possible to have more than one node specified.  For instance, if we wanted to look for genes that link Fanconi Aneima (MONDO:0019391) and DNA repair (GO:0006281), we could rerun our previous query, but setting the final curie, like this:

In [14]:
two_step_question_fixed_ends = \
   make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,'GO:0006281'])
two_step_answer_fixed_ends = quick(two_step_question_fixed_ends)

Return Status: 200


In [15]:
extract_node_names(two_step_answer_fixed_ends)

n0        n1          n2
0   Fanconi anemia     FANCC  DNA repair
1   Fanconi anemia     FANCG  DNA repair
2   Fanconi anemia    RAD51C  DNA repair
3   Fanconi anemia     FANCA  DNA repair
4   Fanconi anemia     FANCL  DNA repair
5   Fanconi anemia      SLX4  DNA repair
6   Fanconi anemia     RAD51  DNA repair
7   Fanconi anemia     ERCC4  DNA repair
8   Fanconi anemia     XRCC2  DNA repair
9   Fanconi anemia     UBE2T  DNA repair
10  Fanconi anemia    FANCD2  DNA repair
11  Fanconi anemia     CENPS  DNA repair
12  Fanconi anemia       XPA  DNA repair
13  Fanconi anemia       WRN  DNA repair
14  Fanconi anemia      FAN1  DNA repair
15  Fanconi anemia    RAD51D  DNA repair
16  Fanconi anemia     SLX1A  DNA repair
17  Fanconi anemia     SLX1B  DNA repair
18  Fanconi anemia       BLM  DNA repair
19  Fanconi anemia     DDX11  DNA repair
20  Fanconi anemia  RAD51AP1  DNA repair
21  Fanconi anemia     MUS81  DNA repair
22  Fanconi anemia    BOD1L1  DNA repair

## Non-path queries

So far, we've only looked at linear paths.  But the question format is actually more general than that - we can define a path pattern generally. So for instance, in the above query, we find all genes that are linked to both FA and DNA repair.  But what if we wanted to find entities that are connected to more than two specified entities.  Here is a query-generation function for the star query:

In [16]:
def make_star_question(types,curies,shared_type,rebuild=None):
    """Create a question to find entities of shared_type that are linked to all of the nodes specified in the
    types and curies arrays."""
    question = {
                #'rebuild': 'True',
                'machine_question': {
                    'nodes': [],
                    'edges': []
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild']='True'
    question['machine_question']['nodes'].append( {'id': 'n0', 'type': shared_type})
    for i,t in enumerate(types):
        newnode = {'id': f'n{i+1}', 'type': t}
        if curies[i] is not None:
            newnode['curie'] = curies[i]
        question['machine_question']['nodes'].append(newnode)
        question['machine_question']['edges'].append( {'id': f'e{i}', 'source_id': 'n0', 'target_id': f'n{i+1}'})
    return question

Suppose I have this set of GO terms, and I'd like to find genes that they all have in common:

* 'voltage-gated sodium channel activity(GO:0005248)',
* 'muscle contraction(GO:0006936)',
* 'voltage-gated ion channel activity(GO:0005244)',
* 'regulation of ion transmembrane transport(GO:0034765)',
* 'sodium ion transmembrane transport(GO:0035725)',
* 'neuronal action potential(GO:0019228)',
* 'membrane depolarization during action potential(GO:0086010)',
* 'sodium ion transport(GO:0006814)'

In [17]:
go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814']
types = ['biological_process_or_activity' for g in go_terms]
star_q = make_star_question(types,go_terms,'gene')
star_q

{'machine_question': {'edges': [{'id': 'e0',
    'source_id': 'n0',
    'target_id': 'n1'},
   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},
   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},
   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},
   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},
   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},
   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},
   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'}],
  'nodes': [{'id': 'n0', 'type': 'gene'},
   {'curie': 'GO:0005248',
    'id': 'n1',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006936',
    'id': 'n2',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0005244',
    'id': 'n3',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0034765',
    'id': 'n4',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0035725',
    'id': 'n5',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0019228',
   

In [18]:
common_gene_answer = quick(star_q)

Return Status: 200


In [19]:
extract_node_names(common_gene_answer)

n0          n1                  n2                                  n3  \
0  SCN4A  GO:0005248  muscle contraction  voltage-gated ion channel activity   
1  SCN7A  GO:0005248  muscle contraction  voltage-gated ion channel activity   

                                          n4  \
0  regulation of ion transmembrane transport   
1  regulation of ion transmembrane transport   

                                   n5                         n6  \
0  sodium ion transmembrane transport  neuronal action potential   
1  sodium ion transmembrane transport  neuronal action potential   

                                                n7                    n8  
0  membrane depolarization during action potential  sodium ion transport  
1  membrane depolarization during action potential  sodium ion transport

We can also have more than one unspecified node.  Suppose, for instance that we wanted to do the previous query, but we also wanted to know what chemicals interact with the genes that we find.  We can do another star query, where one of our spokes is unspecified:

In [20]:
go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814',None]
types = ['biological_process_or_activity' for i in range(8)]+['chemical_substance']
star_q_compound = make_star_question(types,go_terms,'gene',rebuild=True)
star_q_compound

{'machine_question': {'edges': [{'id': 'e0',
    'source_id': 'n0',
    'target_id': 'n1'},
   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},
   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},
   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},
   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},
   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},
   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},
   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'},
   {'id': 'e8', 'source_id': 'n0', 'target_id': 'n9'}],
  'nodes': [{'id': 'n0', 'type': 'gene'},
   {'curie': 'GO:0005248',
    'id': 'n1',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006936',
    'id': 'n2',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0005244',
    'id': 'n3',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0034765',
    'id': 'n4',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0035725',
    'id': 'n5',
    'type': 'biologica

**WARNING: This takes a long time to run**

In [23]:
common_gene_compound_answer = quick(star_q_compound)

Return Status: 200


In [24]:
cgc_nodes = extract_node_names(common_gene_compound_answer)
cgc_nodes[['n0','n9']]

n0                                                 n9
0    SCN4A                                           m-cresol
1    SCN4A                        4-chloro-3,5-dimethylphenol
2    SCN4A                                     4-chlorophenol
3    SCN7A                                           ethotoin
4    SCN7A                                         benzocaine
5    SCN7A                                          tocainide
6    SCN4A                                       pilsicainide
7    SCN4A                                        lamotrigine
8    SCN4A                                           ethotoin
9    SCN7A                                    phenazopyridine
10   SCN4A                         procainamide hydrochloride
11   SCN4A                                          phenytoin
12   SCN7A                         procainamide hydrochloride
13   SCN4A                                         flecainide
14   SCN4A          levobupivacaine hydrochloride (anhydrous)
15   SCN4A                                    (S)-ropivacaine
16   SCN4A                                          lidocaine
17   SCN4A                                         etidocaine
18   SCN4A                                        fospropofol
19   SCN4A                                         benzocaine
20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...
21   SCN4A                                          tocainide
22   SCN4A                                         zonisamide
23   SCN7A                                         rufinamide
24   SCN4A                                    phenazopyridine
25   SCN4A                                        mepivacaine
26   SCN7A                                        mephenytoin
27   SCN7A                                         sodium(1+)
28   SCN4A                                         mexiletine
29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea
..     ...                                                ...
140  SCN4A                            eslicarbazepine acetate
141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine
142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine
143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine
144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...
145  SCN4A                                                 NA
146  SCN4A                                                 NA
147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...
148  SCN4A                                                 NA
149  SCN4A                                                 NA
150  SCN4A                                                 NA
151  SCN4A                                                 NA
152  SCN4A                                                 NA
153  SCN4A                                                 NA
154  SCN4A                                                 NA
155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...
156  SCN4A                                       orphenadrine
157  SCN4A                                                 NA
158  SCN4A                                                 NA
159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine
160  SCN4A                                                 NA
161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole
162  SCN4A                                                 NA
163  SCN4A                                            HWTX-IV
164  SCN4A                                                 NA
165  SCN4A                                                 NA
166  SCN4A                                                 NA
167  SCN4A                                                 NA
168  SCN4A                                                 NA
169  SCN4A                                                 NA

[170 rows x 2 columns]

Many other graph shapes are possible, these are simply a couple of examples.  For instance, we could add onto this query an edge from node 0 to a new disease node (node 10) and also include an edge from the target node (node 9) to this same disease node. Now we would be finding genes that share a set of GO terms, then looking for disease,drug pairs such that the disease, the drug, and the gene are all interconnected.

### Controlling the number of results

By default, quick will return up to 250 results, but this can be raised or lowered using the max_results parameter.  We've already rebuilt this question, so we won't do rebuild=True this time, which should make it go faster.

In [27]:
star_q_compound = make_star_question(types,go_terms,'gene')
common_gene_compound_answer_10 = quick(star_q_compound,max_results=10)

Return Status: 200


In [28]:
cgc_nodes_10 = extract_node_names(common_gene_compound_answer_10)
cgc_nodes_10[['n0','n9']]

n0                                                 n9
0    SCN4A                                           m-cresol
1    SCN4A                        4-chloro-3,5-dimethylphenol
2    SCN4A                                     4-chlorophenol
3    SCN7A                                           ethotoin
4    SCN7A                                         benzocaine
5    SCN7A                                          tocainide
6    SCN4A                                       pilsicainide
7    SCN4A                                        lamotrigine
8    SCN4A                                           ethotoin
9    SCN7A                                    phenazopyridine
10   SCN4A                         procainamide hydrochloride
11   SCN4A                                          phenytoin
12   SCN7A                         procainamide hydrochloride
13   SCN4A                                         flecainide
14   SCN4A          levobupivacaine hydrochloride (anhydrous)
15   SCN4A                                    (S)-ropivacaine
16   SCN4A                                          lidocaine
17   SCN4A                                         etidocaine
18   SCN4A                                        fospropofol
19   SCN4A                                         benzocaine
20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...
21   SCN4A                                          tocainide
22   SCN4A                                         zonisamide
23   SCN7A                                         rufinamide
24   SCN4A                                    phenazopyridine
25   SCN4A                                        mepivacaine
26   SCN7A                                        mephenytoin
27   SCN7A                                         sodium(1+)
28   SCN4A                                         mexiletine
29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea
..     ...                                                ...
140  SCN4A                            eslicarbazepine acetate
141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine
142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine
143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine
144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...
145  SCN4A                                                 NA
146  SCN4A                                                 NA
147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...
148  SCN4A                                                 NA
149  SCN4A                                                 NA
150  SCN4A                                                 NA
151  SCN4A                                                 NA
152  SCN4A                                                 NA
153  SCN4A                                                 NA
154  SCN4A                                                 NA
155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...
156  SCN4A                                       orphenadrine
157  SCN4A                                                 NA
158  SCN4A                                                 NA
159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine
160  SCN4A                                                 NA
161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole
162  SCN4A                                                 NA
163  SCN4A                                            HWTX-IV
164  SCN4A                                                 NA
165  SCN4A                                                 NA
166  SCN4A                                                 NA
167  SCN4A                                                 NA
168  SCN4A                                                 NA
169  SCN4A                                                 NA

[170 rows x 2 columns]

## Rebuild and caching

For details on ROBOKOP caching, and guidance for when to use the `rebuild` parameter, see the notebook on the expand service.  Briefly, the quick service will only run against the cache unless the rebuild parameter is passed, in which case services are re-queried.   The rebuild parameter is passed as part of the json query as seen in the `make_star_query` function above. 

## Edge Properties

In the expand notebook, we saw how a particular predicate could be chosen to limit the results.  This same type information can in fact be added to any edge in a quick query, providing fine control to the results returned.